File: apps/BuddyPlay/android/core/src/main/java/com/americangroupllc/buddyplay/core/storage/LocalRivalryStore.kt
package com.americangroupllc.buddyplay.core.storage

import com.americangroupllc.buddyplay.core.models.GameKind
import com.americangroupllc.buddyplay.core.models.Rivalry
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.json.Json
import kotlinx.serialization.serializer
import java.io.File

/**
 * JSON-on-disk store for cumulative head-to-head records. Single file,
 * single process — no need for App Groups or atomic file replace.
 *
 * JVM-friendly so it can be unit-tested without an Android runtime.
 * `:app` provides the actual `directory` (the app's filesDir).
 */
class LocalRivalryStore(private val directory: File) {

    private val file: File = File(directory, "rivalries.json")
    private val json: Json = Json { encodeDefaults = true; ignoreUnknownKeys = true }
    private val lock = Any()

    init { directory.mkdirs() }

    fun loadAll(): List<Rivalry> = synchronized(lock) {
        if (!file.exists()) return emptyList()
        return try {
            json.decodeFromString(ListSerializer(serializer<Rivalry>()), file.readText())
        } catch (_: Throwable) {
            // Corrupt JSON falls back to empty so a single bad write can
            // never brick the screen.
            emptyList()
        }
    }

    fun load(opponentId: String): Rivalry? = loadAll().firstOrNull { it.opponentId == opponentId }

    fun record(
        opponentId: String,
        opponentName: String,
        kind: GameKind,
        outcome: Rivalry.Outcome,
        atMillis: Long = System.currentTimeMillis(),
    ) = synchronized(lock) {
        val all = loadAll().toMutableList()
        val idx = all.indexOfFirst { it.opponentId == opponentId }
        if (idx >= 0) {
            val r = all[idx]
            r.opponentName = opponentName
            r.record(outcome, kind, atMillis)
        } else {
            val r = Rivalry(opponentId = opponentId, opponentName = opponentName, lastPlayedAt = atMillis)
            r.record(outcome, kind, atMillis)
            all += r
        }
        save(all)
    }

    fun eraseAll() = synchronized(lock) {
        file.delete()
    }

    private fun save(list: List<Rivalry>) {
        file.writeText(json.encodeToString(ListSerializer(serializer<Rivalry>()), list))
    }
}

File: apps/BuddyPlay/shared/BuddyCore/Sources/BuddyCore/Storage/LocalRivalryStore.swift
import Foundation

/// JSON-on-disk store for cumulative head-to-head records. Single file,
/// single process — no need for App Groups or atomic file replace.
public final class LocalRivalryStore {

    private let url: URL
    private let fileManager: FileManager
    private let queue = DispatchQueue(label: "buddyplay.rivalrystore")

    public init(directory: URL? = nil, fileManager: FileManager = .default) throws {
        self.fileManager = fileManager
        let base: URL
        if let directory {
            base = directory
        } else {
            base = try fileManager
                .url(for: .documentDirectory, in: .userDomainMask, appropriateFor: nil, create: true)
                .appendingPathComponent("buddyplay", isDirectory: true)
        }
        try fileManager.createDirectory(at: base, withIntermediateDirectories: true)
        self.url = base.appendingPathComponent("rivalries.json")
    }

    /// Load every rivalry. Returns an empty list if the file doesn't yet
    /// exist or if the JSON is corrupt (corruption falls back to empty so a
    /// single bad write can never brick the screen).
    public func loadAll() -> [Rivalry] {
        queue.sync {
            guard let data = try? Data(contentsOf: url) else { return [] }
            do {
                return try JSONDecoder().decode([Rivalry].self, from: data)
            } catch {
                return []
            }
        }
    }

    public func load(opponentId: UUID) -> Rivalry? {
        loadAll().first { $0.opponentId == opponentId }
    }

    /// Record one game outcome. Idempotent over `(opponentId, kind)`: every
    /// call increments tallies by exactly one.
    public func record(
        opponentId: UUID,
        opponentName: String,
        kind: GameKind,
        outcome: Rivalry.Outcome,
        at date: Date = Date()
    ) {
        queue.sync {
            var all = loadInternal()
            if let idx = all.firstIndex(where: { $0.opponentId == opponentId }) {
                var r = all[idx]
                r.opponentName = opponentName  // keep latest display name
                r.record(outcome, for: kind, at: date)
                all[idx] = r
            } else {
                var r = Rivalry(opponentId: opponentId, opponentName: opponentName, lastPlayedAt: date)
                r.record(outcome, for: kind, at: date)
                all.append(r)
            }
            saveInternal(all)
        }
    }

    /// Wipe all rivalries.
    public func eraseAll() {
        queue.sync {
            try? fileManager.removeItem(at: url)
        }
    }

    // MARK: - Internal

    private func loadInternal() -> [Rivalry] {
        guard let data = try? Data(contentsOf: url) else { return [] }
        return (try? JSONDecoder().decode([Rivalry].self, from: data)) ?? []
    }

    private func saveInternal(_ rivalries: [Rivalry]) {
        let data = try? JSONEncoder().encode(rivalries)
        try? data?.write(to: url, options: .atomic)
    }
}
